src/mesh_smoothing.rs

```diff
@@ -0,0 +1,582 @@
+use nalgebra::geometry::Point3;
+
+use crate::convert::cast_u32;
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::vertex_to_vertex_topology;
+
+/// # Mesh relaxation / Laplacian smoothing
+/// Relaxes angles between mesh edges, resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing(geometry: &Geometry, iterations: u8) -> Geometry {
+    laplacian_smoothing_with_anchors(geometry, iterations, &[])
+}
+
+/// # Mesh relaxation / Laplacian smoothing with fixed vertices
+/// Relaxes angles between mesh edges, while keeping some vertices anchored,
+/// resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing_with_anchors(
+    geometry: &Geometry,
+    iterations: u8,
+    fixed_vertex_indices: &[u32],
+) -> Geometry {
+    let vertex_to_vertex_topology = vertex_to_vertex_topology(geometry);
+    let mut geometry_vertices: Vec<Point3<f32>> = Vec::from(geometry.vertices());
+    let mut vertices = geometry_vertices.clone();
+
+    for _ in 0..iterations {
+        for (current_vertex_index, neighbors_indices) in vertex_to_vertex_topology.iter() {
+            if fixed_vertex_indices
+                .iter()
+                .all(|i| *i != cast_u32(*current_vertex_index))
+            {
+                let mut average_position: Point3<f32> = Point3::origin();
+                for neighbor_index in neighbors_indices {
+                    average_position += geometry_vertices[*neighbor_index].coords;
+                }
+                average_position /= neighbors_indices.len() as f32;
+                vertices[*current_vertex_index] = average_position;
+            }
+        }
+        geometry_vertices = vertices.clone();
+    }
+
+    Geometry::from_triangle_faces_with_vertices_and_normals(
+        geometry.triangle_faces_iter().collect(),
+        vertices,
+        geometry.normals().to_vec(),
```

Comment:
What is more correct? Re-use stale normals after relaxation, or force re-compute of sharp normals?

Reply:
neither of the two. we don't know if the mesh was created with sharp normals or not. also we don't know if the same normal is being reused for each face vertex. 
this is a matter of a decision but i think this should compute smooth normals for the result, as this is a smoothing algorithm. adding a `FIXME`